### PR TITLE
Adding benchmarking for new pallet `pallet_election_provider_support_onchain`

### LINKED ISSUE #11111: Benchmark different implementations of `Solver`
All structs that implement `NposSolver` are currently returning a naive `max_block` weight, just because they don't have any better way to estimate them. 

The weight of these implementations are not really runtime dependent, we can and should still leverage benchmarking pipeline. So, I propose the follwoing: 

2. Create a sub-create in `election-provider-support` called `onchain-config`. This will have a `trait WeightInfo`, similar to all other pallets, with 2 functions

```
trait WeightInfo {
  fn phragmen(<input>) -> Weight;
  fn phragmms(<input>) -> Weight;
}
```

4. This is a raw pallet, with nothing in it other than a:

```
trait Config: frame_system::Config { 
  type WeightInfo: WeightInfo;
  // ++ everything else that's in `trait Config` in `onchain.rs` now.
  /// The accuracy used to compute the election:
  type Accuracy: PerThing128;
  /// Something that provides the data for election.
  type DataProvider: ElectionDataProvider<
    AccountId = Self::AccountId,
    BlockNumber = Self::BlockNumber,
  >;
}
```

6. Then you write benchmarks for this pallet, just as you would do for any other pallet, trying to fulfill the aforementioned `trait WeightInfo`. If any further information is needed to conduct the benchmarks accurately, you can pass in more types to `trait Config`.
7. As you can probably guess now, the existing `trait Config` in `onchain.rs` can be removed. Any struct wanting to be a `OnchainExecution` needs to implement to new `onchain_config::Config` pallet config. 

The main thing that I didn't cover is what is `<input>` in the above snippet. These should be the components that dictate the complexity of phragmen, or similar algorithms. I have tried to do this once in the past, and my conclusion was that this is a function of 3 things: 

1. number of voters
2. number of candidates
3. number of edges (votes per voter)

A similar benchmark was done here: https://github.com/paritytech/substrate/blob/master/frame/elections-phragmen/src/benchmarking.rs#L454

For more general info about benchmarking, see: https://docs.substrate.io/v3/runtime/benchmarking/, https://www.shawntabrizi.com/substrate-graph-benchmarks/docs/#/

### PR DESCRIPTION
**Pull request by georgesdib**
Fixes #11111 

Moving the previous `onchain.rs` into a pallet in order to be able to benchmark it.

Adding benchmarks for it.

I have change quite drastically the API for that, this trait will always need to be bounded now.

### Some thoughts:
`DataProvider` is part of the pallet. This means that only one such provider can be given to that pallet for the full blockchain (unless the pallet is instantiable).
Given that this pallet is meant to be used in a variety of cases, including staking but also governance. Don’t we need multiple `DataProvider` to cater for that?
In other words, should the pallet be instantiable (or take `DataProvider` out of the pallet similar to `NposSolver`)?

Polkadot companion: paritytech/polkadot#5211
skip check-dependent-cumulus

Polkadot address: 131dPecTmpTC1p1ofemufqFBJo9vNbV7dkgN7vWwKnaSMkC4

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -88,6 +88,7 @@ members = [
 	"frame/election-provider-support",
 	"frame/election-provider-support/solution-type",
 	"frame/election-provider-support/solution-type/fuzzer",
+	"frame/election-provider-support/onchain",
```
Comment (kianenigma) [inline Cargo.toml:91]:
```suggestion
	"frame/election-provider-support/onchain-benchmarking",
```

this crate only does benchmarking, it should be named such.

File: frame/election-provider-support/onchain/src/benchmarking.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		let d in (T::BenchmarkingConfig::VOTES_PER_VOTER[0]) .. T::BenchmarkingConfig::VOTES_PER_VOTER[1];
+
+		// we don't directly need the data-provider to be populated, but it is just easy to use it.
+		set_up_data_provider::<T, T::DataProvider>(v, t, d, 1_000u64);
+	}: {
+		assert!(OnChainPhragMMS::<T, sp_runtime::Perbill>::elect().is_ok());
+	} verify {
+	}
+
+	impl_benchmark_test_suite!(
```
Comment (kianenigma) [inline frame/election-provider-support/onchain/src/benchmarking.rs:99]: if there's no `verify` block, I think this does nothing :)) I think we don't need this and the verify block.

**Review by kianenigma — COMMENTED** [review]
I really like the overall direction of this, except one minor problem that I want to raise as soon as possible: 

I don't think we should mix the benchmarking pallet into the existing code. It should, instead, but a new crate. The reason for this is that now, if anyone wants to use `BoundedExecution`, they need to import a whole fake pallet into their runtime. This is more bytes of code in the runtime, which is a problem for both parachians and the relay chain (wasm size does matter!). 

Whereas if you make it a separate pallet, that new pallet is only compiled when you run the `runtime-bechmkarks` command, and it spits out just a single file that contains the weight amounts, which is used in the runtime. 

Other than this, I don't see any major issues with this and hopefully next week I will do a good review. 

Thanks for your contributions!

**Comment by georgesdib** [comment]
> I really like the overall direction of this, except one minor problem that I want to raise as soon as possible:
> 
> I don't think we should mix the benchmarking pallet into the existing code. It should, instead, but a new crate. The reason for this is that now, if anyone wants to use `BoundedExecution`, they need to import a whole fake pallet into their runtime. This is more bytes of code in the runtime, which is a problem for both parachians and the relay chain (wasm size does matter!).
> 
> Whereas if you make it a separate pallet, that new pallet is only compiled when you run the `runtime-bechmkarks` command, and it spits out just a single file that contains the weight amounts, which is used in the runtime.
> 
> Other than this, I don't see any major issues with this and hopefully next week I will do a good review.
> 
> Thanks for your contributions!

Can you please check #11149 and let me know if this is a better direction of travel?

**Comment by georgesdib** [comment]
Closing this one as superseded by #11149
